spec/compiler/crystal/tools/unreachable_spec.cr

```diff
@@ -0,0 +1,413 @@
+
+# References
+#
+#   ༓ marks the expected unreachable code to be found
+#
+describe "unreachable" do
+  it "finds top level methods" do
+    assert_unreachable <<-CRYSTAL
+      ༓def foo
+        1
+      end
+
+      def bar
+        2
+      end
+
+      bar
+      CRYSTAL
+  end
+
+  it "finds instance methods" do
+    assert_unreachable <<-CRYSTAL
+      class Foo
+        ༓def foo
+          1
+        end
+
+        def bar
+          2
+        end
+      end
+
+      Foo.new.bar
+      CRYSTAL
+  end
+
+  it "finds class methods" do
+    assert_unreachable <<-CRYSTAL
+      class Foo
+        ༓def self.foo
+          1
+        end
+
+        def self.bar
+          2
+        end
+      end
+
+      Foo.bar
+      CRYSTAL
+  end
+
+  it "finds instance methods in nested types" do
+    assert_unreachable <<-CRYSTAL
+      module Mod
+        class Foo
+          ༓def foo
+            1
+          end
+
+          def bar
+            2
+          end
+        end
+      end
+
+      Mod::Foo.new.bar
+      CRYSTAL
+  end
+
+  it "finds method with free variable" do
+    assert_unreachable <<-CRYSTAL
+      ༓def foo(u : U) forall U
+      end
+
+      def bar(u : U) forall U
+      end
+
+      bar(1)
+      CRYSTAL
+  end
+
+  it "finds yielding methods" do
+    assert_unreachable <<-CRYSTAL
+      ༓def foo
+        yield
+      end
+
+      def bar
+        yield
+      end
+
+      bar {}
+      CRYSTAL
+  end
+
+  it "finds method called from block" do
+    assert_unreachable <<-CRYSTAL
+      ༓def foo
+      end
+
+      def bar
+      end
+
+      def baz
+        yield
+      end
+
+      baz do
+        bar
+      end
+      CRYSTAL
+  end
+
+  it "finds method called from proc" do
+    assert_unreachable <<-CRYSTAL
+      ༓def foo
+      end
+
+      def bar
+      end
+
+      def baz(&proc : ->)
+        proc.call
+      end
+
+      baz do
+        bar
+      end
+      CRYSTAL
+  end
+
+  it "finds methods with proc parameter" do
+    assert_unreachable <<-CRYSTAL
+      ༓def foo(&proc : ->)
+        proc.call
+      end
+
+      def bar(&proc : ->)
+        proc.call
+      end
+
+      bar {}
+      CRYSTAL
+  end
+
+  it "finds shadowed method (1)" do
+    assert_unreachable <<-CRYSTAL
+      ༓def foo
+      end
+
+      ༓def foo
+      end
+
+      ༓def bar
+      end
+
+      def bar
+      end
+
+      bar
+      CRYSTAL
+  end
+
+  it "finds method with `previous_def`" do
+    assert_unreachable <<-CRYSTAL
+      ༓def foo
+      end
+
+      ༓def foo
+        previous_def
+      end
+
+      def bar
+      end
+
+      def bar
+        previous_def
+      end
+
+      bar
+      CRYSTAL
+  end
+
+  it "finds methods called from reachable code" do
+    assert_unreachable <<-CRYSTAL
+      ༓def qux_foo
+      end
+
+      ༓def foo
+        qux_foo
+      end
+
+      def qux_bar
+      end
+
+      def bar
+        qux_bar
+      end
+
+      bar
+      CRYSTAL
+  end
+
+  it "does not finds method with `super`" do
```

Comment:
What's the difference w.r.t. the other tests?
```suggestion
  it "finds method with `super`" do
```

Reply:
None. That's just a leftover fragment.